### PR TITLE
[Paper+1] PyCells for an Open Semiconductor Industry. Sepideh Alassi, Bertram Winter

### PR DESCRIPTION
**Pull request by SepidehAlassi**
Hi, 

I hope merging of the paper was without problem, however I believe the ./make_paper.sh file needs updates since it adds the EuroSciPy 2013 to the header of paper after it is built!!

Please let me know if there was a problem by submission. 

Best,
Sepideh

### CODE REVIEW

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+:institution: ams AG
+
+
+------------------------------------------------
+PyCells for an Open Semiconductor Industry
+------------------------------------------------
+
+.. class:: abstract
+
+	In modern semiconductor industry, automatic generation of parameterized and recurring layout structures plays an important role and should be present as a feature in Electronic Design Automation (EDA)-tools. Currently these layout generators are developed with a proprietary programming language and can be used with a specific EDA-tool. Therefore, the semiconductor companies find the development of the layout generators that can be used in all state of the art EDA-tools which support OpenAccess database appealing. The goal of this project is to develop computationally efficient layout generators with Python (PyCells), for ams AG technologies, that possess all the features of comprehensive layout generators.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: Typos in the abstract:

...In *the* modern semiconductor industry...

... proprietary programming language and can *only* be used ...

What is "ams"?

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	
+	
+.. class:: keywords
+
+   PyCells, Semiconductor, OpenAccess
+
+Introduction
+------------
+The number of companies active in modern semiconductor business is increasing every day which raises the demand for cheaper EDA-tools. However, one of the main steps of designing an integrated circuit is drawing the physical structure, the layout. Thus the EDA-tools should have a feature for automatic generation of parametrized, recurring structures as pre-defined layouts in a library. The layout generators which are presently in use, are developed using a proprietary language, SKILL, and are only usable as a feature of a specific EDA-tool. On the other hand, the second major drawback of using the proprietary layout generators is the lack of interoperability characteristic. Consequently, the semiconductor manufacturers are not able to support their customers in an optimal way. Hence, the use of an open standard for an EDA database also allows chip designers to work on the same chip-design using tools, which often have more affordable prices, from different EDA vendors.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: ... in *the* modern semiconductor business...

Please define "interoperability characteristic"

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+
+Development
+-----------
+
+OpenAceess [OpenAccess]_ has been established as standard for storing design data in the semiconductor industry and builds the foundation for interoperability of EDA-tools. OpenAccess is a file based database and manages logical design data (schematics) as well as artwork data (layout) for manufacturing. The reference implementation for accessing design data within the database is written in C++. The company Ciranova (now acquired by Synopsys) developed a Python wrapper [PythonAPI]_ for the OpenAccess C++ class library with the goal to access and modify design data through Python. Additionally, they provide an integrated development environment [SynopsysPyCellStudio]_ that enables interactive development and debugging of Python code, while the effects on the design data are shown directly in a separate window of the GUI.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: OpenAccess is spelled wrong in the first sentence of this paragraph.

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+In our development, we have used Python API library to write Python codes which generate recurring layout structures based on a given set of parameters. Typical recurring structures are primitive devices that are used in integrated circuit design such as resistors, capacitors and transistors. The electrical properties of the mentioned devices are brought down to geometric dimensions (e. g. width and length) which serve as parameters for the Python scripts. The object oriented nature of Python matches very well with this task as shown in the following example:
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: I do not understand what "brought down to" means in this context. Please clarify?

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+
+A typical CMOS semiconductor process offers n-channel and p-channel transistors. Each of these transistor types can have multiple voltage ratings (e. g. 5V, 20V, 50V 120V), accordingly resulting in different layout structures. In general, transistors have similar gate, source and drain structures, yet they have eventually different doping concentrations for the implant regions of the semiconductor material and different protection structures, which need to be reflected in the layout. Therefore, we have created a set of base classes in Python from which the more specific transistors inherit common structures, while adding only structures dedicated to the type of transistor. This approach reduces duplicate code and makes maintaining the scripts more intuitive when the number of devices increases. Finally the Python scripts are compiled to byte-code and stored inside the OpenAccess library. A plugin of the OpenAccess database ascertains loading the correct script when a parameter of the corresponding device is modified. In the next section, the development of PyCells is explained using a sample high voltage transistor.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: I do not think 'eventually' is the word you are looking for in this paragraph. Perhaps you mean 'extremely' different?

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+A typical CMOS semiconductor process offers n-channel and p-channel transistors. Each of these transistor types can have multiple voltage ratings (e. g. 5V, 20V, 50V 120V), accordingly resulting in different layout structures. In general, transistors have similar gate, source and drain structures, yet they have eventually different doping concentrations for the implant regions of the semiconductor material and different protection structures, which need to be reflected in the layout. Therefore, we have created a set of base classes in Python from which the more specific transistors inherit common structures, while adding only structures dedicated to the type of transistor. This approach reduces duplicate code and makes maintaining the scripts more intuitive when the number of devices increases. Finally the Python scripts are compiled to byte-code and stored inside the OpenAccess library. A plugin of the OpenAccess database ascertains loading the correct script when a parameter of the corresponding device is modified. In the next section, the development of PyCells is explained using a sample high voltage transistor. 
+
+Sample PyCell Development
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: PyCell - > PyCells?

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
++++++++++++++++++++++++++
+The process of developing a PyCells is described in this section step by step using the sample 20V MOS transistor, pmos20t. In order to simplify the development of new layout generators, the architecture shown in Figure :ref:`architecture` has been established.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: "a PyCells" doesn't make sense (plurality agreement needed). Use either "developing a PyCells object" or "developing a PyCell".

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+.. figure:: pmos50.pdf
+   :scale: 35%
+   :align: center
+   
+   Layout with a 50V guard ring and a device multiplier of 2. :label:`pmos50`
+
+Optimization
+++++++++++++
+
+The computation time plays a crucial role in designing integrated circuits. The complexity of the designs and the amount of operations necessary to generate layouts of high voltage devices or the large devices increase the computation time significantly. For this reason, throughout this project the PyCells are developed to be in the most optimized shape possible without losing the clarity of the structure of the codes. For optimizing the codes, data aggregation and optimal use of local variables as well as iterations are taken into account. This has resulted in faster execution of PyCell codes for generating large and complex devices in comparison with execution time of PCells for identical devices.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: I am not sure I understand what you mean by "shape" in this context:

> For this reason, throughout this project the PyCells are developed to be in the
> most optimized shape possible without losing the clarity of the structure of the 
> codes. 

Do you mean geometric shape of the resistors, or some feature of the software design? I think this paragraph needs to be clarified to specifically address what kind of "optimal use of local variables" and "data aggregation" you implemented. As is, the text is too vague to tell either 

- What you did to optimize the code
- or how fast the code now runs.

Consider answering the following question:

What kind of execution time is typical, and what specific changes to the code were most effective in making this speedup possible.

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+   
+Stretching
+----------
+
+Along with the basic features such as specifying parameters for recurring structures, the Python API also offers more enhanced interactive features with the EDA-tool [PythonAPI]_. Stretching, which is one of these features allows changing parameters of an instance interactively. This helps the engineer to fit instances of an analog design into the available space, without affecting the electrical properties of the instance. For example, the capacitance value of a capacitor depends to a large extent on its area. Therefore, the shape can be varied while maintaining a constant area. An example of stretching is shown for a rectangular capacitor in Figure :ref:`capfig`:
+
+.. math::
+   :label: capequ
+
+   C \propto W \cdot L
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: I believe this paragraph will read more smoothly if this equation appears after "area." and before "An example of stretching... "

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+   
+   Stretching of a transistor by increasing the number of fingers, but keeping the gate to source/drain edge constant. :label:`transfig`
+   
+.. figure:: stretch_gate.pdf
+   :scale: 80%
+   :align: center
+   
+   Stretching the gate of a transistor :label:`stretchgate`
+   
+The stretch handles are stored as properties in the OpenAccess database and are defined within the Python source code. The stretch handles have to be associated to the shape and parameters of the layout. Stretching can be limited to minimum and maximum boundaries on both horizontal and vertical directions. Snap resolution can be also be specified in definition of stretch handles, an example of which is shown below:
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: Please be more clear about what you mean by "stretch handles."

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+
+Schematic Driven Layout (SDL)
++++++++++++++++++++++++++++++
+PyCell codes must include the definition of pins regarding type of connections, weak connection (by polysilicon) or strong connection (by metal), and also the definition of shapes requiring external connections. Having types of connections implemented in PyCells, the SDL Navigator tool can attain the schematic of a sample design and generate the layout of it by calling PyCells with respective parameters. After generation of the layout, SDL Navigator checks the connectivity features using flylines.
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: I don't know what a flyline is. Please be more explicit?

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+
+
+Conclusion
+------------
+
+Believing in what Richard Stallman has said "Proprietary software is an injustice", we have successfully developed the layout generators for an open semiconductor industry, giving the designers who use ams AG technologies, opportunity to work with any desired EDA-tools, to share their designs easily with others, and to receive the support they deserve. These PyCells are developed for all high and low voltage devices of ams AG technologies in order to enable the IC-designers to have advanced designs without having constraints because of missing parametrized layouts. Furthermore, by using these PyCells we have managed to reduce the computation time which is extremely important in semiconductor industry. The developed PyCells have passed all the verification tests and in practical use by customers.     
+
+
+References
```
Comment (katyhuff) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: In these references, please point to the location of the software, it's documentation, and its license. You currently point to "synopsis.com" which redirects immediately to a secure login page. This does not seem to be particularly in line with the openness touted in this paper. Is the tool not actually open source?

**Comment by katyhuff** [comment]
Independent Review Report
=========================

.. note::  Please be aware that all reviews are made public including
           the reviewer's name.

Reviewer: Kathryn Huff

Department/Center/Division: Department of Nuclear Engineering and Berkeley Institute for Data Science 

Institution/University/Company: University of California, Berkeley

Field of interest / expertise: Nuclear Engineering, Python, C++

Country: USA

Article reviewed: PyCells for an Open Semiconductor Industry. Sepideh Alassi, Bertram Winter #47

GENERAL EVALUATION
------------------

Please rate the paper using the following criteria (please use the abbreviation
to the right of the description)::

  below     doesn't meet standards for academic publication
  meets     meets or exceeds the standards for academic publication
  n/a       not applicable

- Quality of the approach: meets

- Quality of the writing: meets (minor typos noted and should be corrected)

- Quality of the figures/tables: meets


SPECIFIC EVALUATION
-------------------

For the following questions, please respond with 'yes' or 'no'.  If you
answer 'no',  please provide a brief, one- to two-sentence explanation.

- Is the code made publicly available and does the article sufficiently
  describe how to access it?  We aim not to publish papers that essentially
  advertise proprietary software.  Therefore, if the code is not publicly
  available, please provide a one- to two- sentence response to each of the
  following questions:

**No. It is not clear from the current document where PyCells can be found, how to acquire it, or what its license is. The reference currently in place in the paper redirects to a secure login page.**

  - Does the article focus on a topic other than the features
    of the software itself?

**No.**

  - Can the majority of statements made be externally validated
    (i.e., without the use of the software)?

**No.**

  - Is the information presented of interest to readers other than
    those at whom the software is aimed?

**Yes.**

  - Is there any other aspect of the article that would
    justify including it despite the fact that the code
    isn't available?

**No.**

  - Does the article discuss the reasons the software is closed?

**No. In fact, it indicates the opposite. The introduction seems to hint that the software is open and available. Perhaps this is the case. However, the location of the software or how to access it is not obvious in the article. So, this article may simply need to be updated to reflect the location of the PyCells software.**

- Does the article present the problem in an appropriate context?
  Specifically, does it:

  - explain why the problem is important,
  - describe in which situations it arises,
  - outline relevant previous work,
  - provide background information for non-experts

**Yes.**

- Is the content of the paper accessible to a computational scientist
  with no specific knowledge in the given field?

**Yes. I noted some specific instances in which jargon was confusing. Those instances should be fixed. Once that is done, the paper will be extremely accessible to a computational scientists from a different field.**

- Does the paper describe a well-formulated scientific or technical
  achievement?

**Yes. (the production of a library and improvement of a gui for circuit design)**

- Are the technical and scientific decisions well-motivated and
  clearly explained?

**Yes**

- Are the code examples (if any) sound, clear, and well-written?

**Yes**

- Is the paper factual correct?

**Probably. (location of code not clear, so this cannot be confirmed.)**

- Is the language and grammar of sufficient quality?

**Yes**

- Are the conclusions justified?

**Yes, if the claims about the software performance and functionality are true, then the conclusion (this software should make circuit design easier) is true.**

- Is prior work properly and fully cited?

**I don't think so. The references are very limited. I am not particularly aware of prior work, but some efforts are discussed without being referenced. This should be improved.**

- Should any part of the article be shortened or expanded? Please explain.

**The discussion of optimizing the software should be discussed in more specific detail.**

- In your view, is the paper fit for publication in the conference proceedings?
  Please suggest specific improvements and indicate whether you think the
  article needs a significant rewrite (rather than a minor revision).

**If it turns out that the the software is open, then yes, this is fit for publication and the needed changes are minor. The most important needed change in that case is : list and cite the location of the software, its documentation, and its license.**

**If the software is proprietary, then no, it is not fit for publication.**

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 3 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: thanks for it, already corrected the typo

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 1 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: have added the "the" to the sentence, 
ams AG, formerly known as austriamicrosystems AG and still known as AMS, is a multinational semiconductor manufacturer. we have developed the PyCells for technologies of ams AG, as mentioned,  the coauthor, Bertram Winter, is an employee of this foundry

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 2 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: there are many EDA(Electronic design automation)-tools for circuit design, such as cadence, synopsis, tanner, ...  the purpose of our work was to develop libraries of layout generators for technologies of ams AG, which can be used by different EDA-tools. By interoperability characteristic, we mean that the libraries can be used by various tools without any restriction and designer is free to choose the EDA-tool he/she prefers. Originally all these libraries were developed by SKILL language and could be just used by cadence. The high cost of most of EDA tools was the reason that most of the small companies were not able to afford this tool, and consequently couldn't use the libraries of layout generators.

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 4 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: we mean simplified in this context, I have currently changed the words.

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 5 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: we meant that transistors have same gate, source and drain, but at the end different doping, the sentence has the same meaning without this word, so I erased it.

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 7 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: typo is fixed

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 10 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: Thanks for suggestion, I have added the definition of stretch handles

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 11 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: Thanks for suggestion, I have added the definition of flylines

File: papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry (hunk 12 above)
Comment (SepidehAlassi) [inline papers/Sepideh_Alassi_Bertram_Winter/PyCells_for_an_Open_Semiconductor_Industry]: The tool which we have used to develop the PyCells is called Ciranova which belongs to Synopsis whose products are not open source. This tool is just used by us for development purpose, and the users of our libraries don't need to have or use this development interface in order to execute the codes, PyCells. The PyCell codes can be executable by any EDA-tool which has Python compiler.

**Pull request merged**